File: Src/java/elm-fhir/src/main/java/org/cqframework/cql/elm/requirements/ElmConjunctiveRequirement.java
package org.cqframework.cql.elm.requirements;

import org.hl7.elm.r1.Expression;
import org.hl7.elm.r1.VersionedIdentifier;

import java.util.ArrayList;
import java.util.List;

public class ElmConjunctiveRequirement extends ElmExpressionRequirement {
    public ElmConjunctiveRequirement(VersionedIdentifier libraryIdentifier, Expression expression) {
        super(libraryIdentifier, expression);
    }

    private List<ElmExpressionRequirement> arguments = new ArrayList<ElmExpressionRequirement>();
    public List<ElmExpressionRequirement> getArguments() {
        return arguments;
    }

    @Override
    public ElmExpressionRequirement combine(ElmRequirement requirement) {
        if (requirement instanceof ElmConjunctiveRequirement) {
            for (ElmExpressionRequirement argument : ((ElmConjunctiveRequirement)requirement).getArguments()) {
                arguments.add(argument);
            }
        }
        else if (requirement instanceof ElmDisjunctiveRequirement) {
            // Conjunction of disjunctions, too complex for analysis (i.e. not in DNF)
            return new ElmExpressionRequirement(this.libraryIdentifier, this.getExpression());
        }
        else if (requirement instanceof ElmExpressionRequirement) {
            arguments.add((ElmExpressionRequirement)requirement);
        }
        else if (requirement instanceof ElmRequirements) {
            for (ElmRequirement r : ((ElmRequirements)requirement).getRequirements()) {
                combine(r);
            }
        }
        return this;
    }
}

File: Src/java/elm-fhir/src/main/kotlin/org/cqframework/cql/elm/requirements/ElmConjunctiveRequirement.kt
package org.cqframework.cql.elm.requirements

import org.hl7.elm.r1.Expression
import org.hl7.elm.r1.VersionedIdentifier

class ElmConjunctiveRequirement(libraryIdentifier: VersionedIdentifier, expression: Expression) :
    ElmExpressionRequirement(libraryIdentifier, expression) {
    @JvmField val arguments: MutableList<ElmExpressionRequirement?> = ArrayList()

    @Suppress("ForbiddenComment")
    override fun combine(requirement: ElmRequirement?): ElmExpressionRequirement {
        when (requirement) {
            is ElmConjunctiveRequirement -> {
                for (argument in requirement.arguments) {
                    arguments.add(argument)
                }
            }

            is ElmDisjunctiveRequirement -> {
                // TODO: This should return the disjunctive requirement
                // However, disjunctive requirements can't be distributed without a union rewrite
                // if (arguments.size == 0) {
                //    return requirement;
                // }

                // Conjunction of disjunctions, too complex for analysis (i.e. not in DNF)
                return ElmExpressionRequirement(this.libraryIdentifier, this.expression)
            }

            is ElmExpressionRequirement -> {
                arguments.add(requirement)
            }

            is ElmRequirements -> {
                for (r in requirement.getRequirements()) {
                    combine(r)
                }
            }
        }
        return this
    }
}
